src/State/Hub.php

```diff
@@ -0,0 +1,225 @@
+    }
+
+    /**
+     * Gets the client binded to the top of the stack.
+     *
+     * @return ClientInterface|null
+     */
+    public function getClient(): ?ClientInterface
+    {
+        return $this->getStackTop()->getClient();
+    }
+
+    /**
+     * Gets the scope binded to the top of the stack.
+     *
+     * @return Scope
+     */
+    public function getScope(): Scope
+    {
+        return $this->getStackTop()->getScope();
+    }
+
+    /**
+     * Gets the stack of clients and scopes.
+     *
+     * @return Layer[]
+     */
+    public function getStack(): array
+    {
+        return $this->stack;
+    }
+
+    /**
+     * Gets the topmost client/layer pair in the stack.
+     *
+     * @return Layer
+     */
+    public function getStackTop(): Layer
+    {
+        return $this->stack[\count($this->stack) - 1];
+    }
+
+    /**
+     * Gets the ID of the last captured event.
+     *
+     * @return null|string
+     */
+    public function getLastEventId(): ?string
+    {
+        return $this->lastEventId;
+    }
+
+    /**
+     * Creates a new scope to store context information that will be layered on
+     * top of the current one. It is isolated, i.e. all breadcrumbs and context
+     * information added to this scope will be removed once the scope ends. Be
+     * sure to always remove this scope with {@see Hub::popScope} when the
+     * operation finishes or throws.
+     *
+     * @return Scope
+     */
+    public function pushScope(): Scope
+    {
+        $clonedScope = clone $this->getScope();
+
+        $this->stack[] = new Layer($this->getClient(), $clonedScope);
+
+        return $clonedScope;
+    }
+
+    /**
+     * Removes a previously pushed scope from the stack. This restores the state
+     * before the scope was pushed. All breadcrumbs and context information added
+     * since the last call to {@see Hub::pushScope} are discarded.
+     *
+     * @return bool
+     */
+    public function popScope(): bool
+    {
+        if (1 === \count($this->stack)) {
+            return false;
+        }
+
+        return null !== \array_pop($this->stack);
+    }
+
+    /**
+     * Creates a new scope with and executes the given operation within. The scope
+     * is automatically removed once the operation finishes or throws.
+     *
+     * @param callable $callback The callback to be executed
+     */
+    public function withScope(callable $callback): void
+    {
+        $scope = $this->pushScope();
+
+        try {
+            $callback($scope);
+        } finally {
+            $this->popScope();
+        }
+    }
+
+    /**
+     * Calls the given callback passing to it the current scope so that any
+     * operation can be run within its context.
+     *
+     * @param callable $callback The callback to be executed
+     */
+    public function configureScope(callable $callback): void
+    {
+        $layer = $this->getStackTop();
+
+        if (null !== $layer->getClient()) {
```

Comment:
Because if no client is set on the hub, it should be a noop.

Reply:
I'm not sure, I just ported the code from the Javascript SDK which does the following check:

```js
if (top.scope && top.client) {
   ...
}
```

Actually the first check is not needed because there will always be a scope (or you pass it explicitly when creating a nwe `Hub` instance or it will be created for you internally in the constructor so the other check left was the one on the client. Maybe @HazAT can explain us why it's needed, I see no reason regarding the code as we never use the client so I agree that we shouldn't care about it